react/components/GoogleMap.tsx

```diff
@@ -0,0 +1,376 @@
+import type { ReactNode, CSSProperties } from 'react'
+import React, {
+  useImperativeHandle,
+  forwardRef,
+  useRef,
+  useMemo,
+  useCallback,
+  createContext,
+  useState,
+  useEffect,
+  useContext,
+} from 'react'
+
+import { useModalState } from '../modalStateContext'
+import { LIST } from '../constants'
+
+interface Context {
+  map: google.maps.Map | null
+  onZoomIn: () => void
+  onZoomOut: () => void
+}
+
+const ctx = createContext<Context | null>(null)
+
+export const useMaps = () => {
+  const contextValue = useContext(ctx)
+
+  if (contextValue == null) {
+    throw new Error('useMaps must be used inside GoogleMaps component')
+  }
+
+  return contextValue
+}
+
+interface Props {
+  /**
+   * Center of the map in the format of [longitude, latitude]
+   */
+  center?: [number, number]
+  /**
+   * Wheter or not we are in a large screen (desktop). This is useful for the
+   * recentering logic to accomodate for the extra UI displayed on top of the map
+   */
+  isLargeScreen?: boolean
+  /**
+   *
+   * Whether or not we are still waiting for the Google Maps SDK to load
+   */
+  isLoading: boolean
+  /**
+   * Element to display while loading the Google Maps SDK
+   */
+  loadingElement: ReactNode
+  /**
+   * Click handler for when the map itself is clicked
+   */
+  onClick?: () => void
+  /**
+   * Reference point in map to use for when prompting to search another area
+   * besides the surroundings of the address geo coordinates
+   */
+  referenceCenter?: [number, number]
+  /**
+   * Bounds of the map
+   * @see {@type google.maps.Map['fitBounds']}
+   */
+  bounds?: google.maps.LatLngBounds | null
+}
+
+/**
+ * Distance of map center to start prompting to search area
+ */
+const DISTANCE = 50
+
+const STANDARD_ZOOM = 14
+
+/**
+ * Horizontal panning to accomodate the extra UI in large screns
+ */
+const PAN_LEFT_LAT = -160
+/**
+ * Vertical panning to accomodate the extra UI in large screns
+ */
+const PAN_LEFT_LNG = -30
+
+const MAP_STYLES: CSSProperties = {
+  height: '100%',
+  width: '100%',
+  position: 'absolute',
+  top: 0,
+  zIndex: 0,
+}
```

Comment:
Nice way of using inline styles and arguably have a better DX than using standard CSS files (for this case, at least) ✨